First Pass

mcdonnelldean opened the pull request:
Just some tidy of tests and examples,

I've more to add but I'd like to get the structure synced so I don't hit rebase hell later. Don't pull this yet as there is still docs to finish before I am happy.
- added npm debug logs to .gitignore
- moved tests and fixtures into /test, update package.json accordingly
- moved example.js into /eg
- added emit before on example
- added idempotency example
- added example schema for example files
- added basic-example.js
- tidied over some naming and comments
- Updated Docs

mcdonnelldean committed:
First Pass
- added npm debug logs to .gitignore
- moved tests and fixtures into /test, update package.json accordingly
- moved example.js into /eg
- added emit before on example
- added idempotency example
- added example schema for example files

mcdonnelldean committed:
Spelling issues and minor comment tidy

mcdonnelldean committed:
more examples
- added basic-example.js
- tidied over some naming and comments

mcdonnelldean committed:
More examples
- added multiple-connected-emitters.js example
- removed old example file
- fixed minor comment issues

mcollina commented on eg/basic-example.js:
```diff
@@ -0,0 +1,67 @@
+// A bare minimum example of two subscribers
+// handling some messages.
+
+'use strict'
```
This repo is not using 'use strict' everywhere, however I am ok to add it. In case, place it everywhere.

mcdonnelldean replied:
Cool, I'll add as move around.

mcollina commented on eg/emit-before-on.js:
```diff
@@ -0,0 +1,44 @@
+  emitter.on('userAdded', function (msg) {
+    console.log('userAdded: ', msg)
+    emitter.close()
+  })
```
Technically, this happens because `emit` is not a synchronous operation anymore. So when you call `emit` and `on`, you actually subscribing _before_ the data is emitted.

In order to demonstrate the goal of the example, you should use two HyperEmitter, and seeing that the second is receiving messages emitted by the first.

mcdonnelldean replied:
@mcollina So if i put a long enough delay in before calling on then I will not get the message as expected?

mcdonnelldean replied:
Ah I see what you mean, this sample is not correct because if I do this;

```
  emitter.emit('userAdded: ', userAddedMsg, function () {
    emitter.on('userAdded', function (msg) {
        console.log('userAdded: ', msg)
        emitter.close()
    })
  })
```

It does not work, where as my sample claims it does. Cool, I can change this sample.

So in other for the emit to actually happen there needs to be at least one listener subscribed?

mcollina commented on eg/message-idempotency.js:
```diff
@@ -0,0 +1,73 @@
+  emitter.emit('userAdded', userAddedMsg)
+  emitter.emit('userAdded', userAddedMsg)
+  emitter.emit('userAdded', userAddedMsgDup)
+  emitter.emit('userAdded', userAddedMsgAdditional)
```
This works because these are all emitted at the same time, if you chain them you will get a different result:

``` js
emitter.emit('userAdded', userAddedMsg, function () {
  emitter.emit('userAdded', userAddedMsg, function () {
    emitter.emit('userAdded', userAddedMsgDup, function () {
      emitter.emit('userAdded', userAddedMsgAdditional, function () {
        console.log('all completed')
      })
    })
  })
})
```

While I think idempotency is a cool feature, it is technically not working in all cases on multiple hosts. Even if we can make it work in a distributed manner,  I fear this has some obscure performance result during replication. 
So, this is kind of a side effect, and I would prefer not to showcase it. Or, if we want to showcase it, we should tell the differences.

mcdonnelldean replied:
@mcollina Interesting, I was not expecting that. So what is happening here. Is it that messages are getting lost or how is it that waiting for the callback changes this?

mcollina replied:
The key is the `_last` property, that contains the key to the last element added to the store by this peer. This key is added in the hash, and it's important for making synchronization fast (basically each child validates its parents).
As you can see here: https://github.com/mcollina/hyperemitter/blob/master/hyperemitter.js#L130, if you do two subsequent emits, those will have the same `_last` value, and if they have the same content they will have the same hash. Does that makes any sense?

The question is what we want to do with this problem/feature, should we leave it in, or we make sure every of them as a different hash? This can be done by adding a counter to the event. Or I'm listening to other ideas. However, does it needs fixing?

mcollina commented:
Thanks for the help, it is really welcomed! I hope the various comments clarifies things a bit.

mcdonnelldean commented:
@mcollina No hassle :D I think examples help. Don't pull it until the examples a technically correct though. I want them to be right.

mcdonnelldean committed:
Added more correct examples, removed bad ones

mcdonnelldean committed:
Documation Goodness
- Broke out readme
- Added documentation basics

mcdonnelldean committed:
Added Contributing Doc

mcdonnelldean committed:
added api ref doc

mcdonnelldean committed:
Moved across CLI doc

mcdonnelldean commented:
@mcollina Ok, this is me finished on this PR, let me know what needs changing.

mcollina commented:
I'm not in for having a separate folder for the docs, can you please get them back into the main README file? This module should be simple enough to a have a single page README, and I have some thoughts that this module is already doing too much :D. I also prefer the examples folder be named `examples`.
I would rename the `docs/contributing.md` file into `CONTRIBUTING.md`.

mcdonnelldean commented:
Cool. Let me change in around in an hour or so and re push

mcdonnelldean committed:
Implemented Matteo's suggested changes
- renamed /eg to /examples
- merged some docs back to readme
- added CONTRIBUTING file
- removed line 111 from hyperemitter, causing lint issues

mcdonnelldean commented:
@mcollina Take a gander over that.

mcdonnelldean commented:
@mcollina IO.js is giving the issue below But I'm not getting the issue with IO.js 2.0.0

mcollina commented:
leveldown requires a bump, it's coming tomorrow or earlier
Il giorno mar 5 mag 2015 alle 12:09 Dean McDonnell notifications@github.com
ha scritto:

> @mcollina https://github.com/mcollina ^^ IO.js is giving the issue But
> I'm not getting the issue with IO.js 2.0.0
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/mcollina/hyperemitter/pull/1#issuecomment-99018623.

mcdonnelldean committed:
re-added example

mcdonnelldean commented:
@mcollina Cool, interersting, I presume it worked locally because I already have level built? Not to worry, I have added back the example too.
